File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.c
#include <stdio.h>
#include <stdbool.h>
#define BUFSIZE 1 << 15
char readbuf[BUFSIZE], writebuf[BUFSIZE];
int rp = BUFSIZE, wp = 0;
char ReadChar(){
    if(rp == BUFSIZE){
        fread(readbuf, 1, BUFSIZE, stdin);
        rp = 0;
    }
    return readbuf[rp++];
}
void Write(bool tf){
    if(wp + 5 >= BUFSIZE){
        fwrite(writebuf, 1, wp, stdout);
    }
    if(tf){
        writebuf[wp++] = 'y';
        writebuf[wp++] = 'e';
        writebuf[wp++] = 's';
        writebuf[wp++] = '\n';
    }else{
        writebuf[wp++] = 'n';
        writebuf[wp++] = 'o';
        writebuf[wp++] = '\n';
    }
}

char stack[102];
int top = -1;

void push(char data){
    stack[++top] = data;
}
int pop(){
    return stack[top--];
}

int main(void){
    char c;
    while(true){
        c = ReadChar();
        bool fail = false;
        top = -1;
        if(c == '.') break;

        while(c != '.'){
            if(fail){
                c = ReadChar();
                continue;
            }
            if(c == '(' || c == '[') push(c);
            else if(c == ')' || c == ']') {
                if(top == -1) fail = true;
                else{
                    if(stack[top] == '(' && c == ')') pop();
                    else if(stack[top] == '[' && c == ']') pop();
                    else fail = true;
                }
            }
            c = ReadChar();
        }
        if(top != -1 || fail) Write(false);
        else Write(true);

        c = ReadChar(); // LF
    }
    fwrite(writebuf, 1, wp, stdout);
}

File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.Stack;
import java.util.StringTokenizer;

public class Main {
    private char[] stack = new char[100001];
    private int top = -1;

    public void push(char data){
        stack[++top] = data;
    }
    public char pop(){
        return stack[top--];
    }

    public void solution() throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        // Stack<Integer> stack = new Stack<>();
        
        while(true){
            String line = br.readLine();
            if(".".equals(line)) break;

            char[] str = new char[102];
            str = line.toCharArray();
            
            top = -1;
            boolean fail = false;
            int length = line.length();
            for(int i = 0; i < length; i++){
                if(fail) break;
                char c = str[i];

                if(c == '(' || c == '[') push(c);
                else if(c == ')' || c == ']'){
                    if(top == -1) fail = true;
                    else{
                        if(stack[top] == '(' && c == ')') pop();
                        else if(stack[top] == '[' && c == ']') pop();
                        else fail = true;
                    }
                }
            }
            if(top != -1 || fail) bw.write("no" + '\n');
            else bw.write("yes" + '\n');
        }

        bw.flush();
        bw.close();
        br.close();
    }
    public static void main(String[] args) throws Exception {
        new Main().solution();
    }
}
